posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,631 @@
+import json
+import os
+from collections.abc import Iterator
+from dataclasses import asdict, dataclass, field
+from datetime import datetime, timedelta, timezone
+from typing import Iterable, NamedTuple
+from uuid import UUID
+
+import psycopg2
+from django.conf import settings
+from temporalio import activity, workflow
+from temporalio.common import RetryPolicy
+
+from posthog.clickhouse.client.execute import sync_execute
+from posthog.temporal.workflows.base import CommandableWorkflow
+
+EPOCH = datetime(1970, 1, 1, 0, 0, tzinfo=timezone.utc)
+
+SELECT_PERSONS_TO_DELETE_QUERY = """
+SELECT
+    team_id,
+    old_person_id,
+    override_person_id,
+    max(created_at) AS latest_created_at,
+    max(version) AS latest_version,
+    min(oldest_event) AS oldest_event_at
+FROM
+    {database}.person_overrides
+WHERE
+    created_at <= %(latest_created_at)s
+GROUP BY
+    team_id, old_person_id, override_person_id
+"""
+
+SELECT_LATEST_CREATED_AT_QUERY = """
+SELECT
+    max(created_at)
+FROM {database}.person_overrides;
+"""
+
+CREATE_JOIN_TABLE_QUERY = """
+CREATE OR REPLACE TABLE {database}.{table_name} (
+    team_id INT NOT NULL,
+    old_person_id UUID NOT NULL,
+    override_person_id UUID NOT NULL
+)
+ENGINE = Join(ANY, LEFT, team_id, old_person_id);
+"""
+
+INSERT_INTO_JOIN_TABLE_QUERY = """
+INSERT INTO {database}.{table_name}
+    SELECT
+        team_id,
+        old_person_id,
+        argMax(override_person_id, created_at) AS override_person_id
+    FROM {database}.person_overrides
+    WHERE created_at <= %(latest_created_at)s
+    GROUP BY team_id, old_person_id;
+"""
+
+SQUASH_EVENTS_QUERY = """
+ALTER TABLE
+    {database}.sharded_events
+UPDATE
+    person_id = joinGet({database}.{table_name}, 'override_person_id', toInt32(team_id), person_id)
+IN PARTITION
+    %(partition_id)s
+WHERE
+    true;
+"""
+
+DROP_JOIN_TABLE_QUERY = """
+DROP TABLE {database}.{table_name};
+"""
+
+DELETE_SQUASHED_PERSON_OVERRIDES_QUERY = """
+ALTER TABLE
+    {database}.person_overrides
+DELETE WHERE
+    team_id = %(team_id)s
+    AND old_person_id = %(old_person_id)s
```

Comment:
we should not do one by one deletes, even in a tiny table like this - why not use just `merged_at`? I understand we might fail to write to the other table or something of that nature but yeah, we definitely need to find a way to batch deletes. 

(or potentially as an alternative we can upgrade CH first, see https://clickhouse.com/docs/en/guides/developer/lightweght-delete)

Reply:
We do use `merged_at`: it's the last predicate.

We could try IN checks, e.g.
```
DELETE WHERE
    old_person_id IN (...)
```

Since `old_person_id` is a UUID there is no risk of collision between teams (and we don't need `team_id`).

Reply:
Gave this a shot, tests seem to pass.